src/data/articles/compile-time-data-contracts-scala-3/index.mdx

```diff
@@ -0,0 +1,1732 @@
+      val sh = Expr.summon[Shape[A]].getOrElse(report.errorAndAbort("No Shape available"))
+      '{
+        val fs = $sh.fields
+        fs.map(f => (f.name, f.tpe, f.hasDefault, f.isOptional))
+      }.valueOrAbort
+
+    val out      = shapeOf[Out]
+    val contract = shapeOf[Contract]
+
+    // Compare fields by name (unordered, case-sensitive for now)
+    val outMap      = out.map { case (n,t,_,opt) => n -> (t,opt) }.toMap
+    val contractMap = contract.map { case (n,t,_,opt) => n -> (t,opt) }.toMap
+
+    val missing = contract.collect { case (n,t,_,opt) if !outMap.contains(n) => s"$n:$t${if opt then " (optional)" else ""}" }
+    val extra   = out.collect      { case (n,_,_,_) if !contractMap.contains(n) => n }
+    val mism    = contract.collect {
+      case (n,t,_,opt) if outMap.get(n).exists(_._1 != t)  => s"$n expected $t, found ${outMap(n)._1}"
+      case (n,_,_,opt) if outMap.get(n).exists(_._2 != opt) => s"$n optionality mismatch"
+    }
+
+    val (okMissing, okExtra, okMism) = Type.of[P] match
+      case '[Policy.Exact]    => (missing, extra, mism)
+      case '[Policy.Backward] => (missing.filterNot(_.contains("(optional)")), Nil, mism)
+      case '[Policy.Forward]  => (Nil, extra, mism)
+      case _                  => (missing, extra, mism)
+
+    if okMissing.nonEmpty || okExtra.nonEmpty || okMism.nonEmpty then
+      val msg = s"""
+        |Compile‑time contract drift (policy: ${Type.show[P]}).
+        |Out: ${Type.show[Out]} vs Contract: ${Type.show[Contract]}
+        |Missing: ${okMissing.mkString(", ")}
+        |Extra: ${okExtra.mkString(", ")}
+        |Mismatched: ${okMism.mkString("; ")}
+        |""".stripMargin
+      quotes.reflect.report.errorAndAbort(msg)
+    else '{ new Conforms[Out, Contract, P] }
+```
+
+Let me break down what's happening:
+
+1. **Extract shapes**: We summon `Shape[Out]` and `Shape[Contract]` and get their field lists. Note the `.valueOrAbort` - this forces compile-time evaluation.
+
+2. **Compare**: We build maps of field names → (type, optionality) and find differences.
+
+3. **Apply policy**:
+   - Exact: all differences are errors
+   - Backward: missing optional fields are OK, extras are OK
+   - Forward: extras are OK (but missing required fields are not)
+
+4. **Abort or succeed**: If there are violations, we call `report.errorAndAbort` with a detailed message showing exactly what's wrong. Otherwise, we return the evidence.
+
+The beauty here: **this runs during compilation**. If schemas don't match, your code won't compile. Period.
+
+:::danger[Build breakage alert]
+Compile-time contract checks will break your build when schemas drift. That's the point, but coordinate with your team. When upstream adds a field, every downstream pipeline build fails until they update contracts or transforms. This forces alignment, but can block deployments if not managed. Use CI branch builds to preview impact before merging schema changes.
+:::
+
+Use it like this:
```

Comment:
I think it's worth explaining this _just a bit more_. The drift part is important to see in practice. The reader might think that you're just adding a field to your case class (or something similar), so what are you comparing to? The example below is obvious but a bit contrived. Can you give a more practical one?